### PR TITLE
Handling local_to_global() IndexError

### PR DESCRIPTION
**Pull request by kywch**

### CODE REVIEW

File: pokemonred_puffer/environment.py [hunk 1]
```diff
@@ -818,7 +822,7 @@ def get_explore_map(self):
         explore_map = np.zeros(GLOBAL_MAP_SHAPE)
         for (x, y, map_n), v in self.seen_coords.items():
             gy, gx = local_to_global(y, x, map_n)
-            if gy >= explore_map.shape[0] or gx >= explore_map.shape[1]:
+            if gy >= explore_map.shape[0] or gy < 0 or gx >= explore_map.shape[1] or gx < 0:
```
Comment (thatguy11325) [inline pokemonred_puffer/environment.py]:
```suggestion
            if 0 > gy >= explore_map.shape[0] or  0 > gx >= explore_map.shape[1]:
```

Yes you can do this in Python :)

File: pokemonred_puffer/global_map.py [hunk 2]
```diff
@@ -20,3 +20,6 @@ def local_to_global(r: int, c: int, map_n: int):
     except KeyError:
         print(f"Map id {map_n} not found in map_data.json.")
         return r + 0, c + 0
+    except IndexError:
+        print(f"Coord {map_x}, {map_y} out of bounds for map id {map_n}.")
+        return -1, -1
```
Comment (thatguy11325) [inline pokemonred_puffer/global_map.py]: Wont this index to the lower right corner of the map? Why not use 0, 0?

File: pokemonred_puffer/environment.py [hunk 3]
```diff
@@ -309,7 +309,7 @@ def init_mem(self):
         self.seen_hidden_objs = {}
 
         self.cut_coords = {}
-        self.cut_tiles = set([])
+        self.cut_tiles = {}  # set([])
```
Comment (thatguy11325) [inline pokemonred_puffer/environment.py]:
```suggestion
        self.cut_tiles = {}
```

File: pokemonred_puffer/environment.py [hunk 4]
```diff
@@ -809,7 +809,11 @@ def update_seen_coords(self):
         x_pos, y_pos, map_n = self.get_game_coords()
         self.seen_coords[(x_pos, y_pos, map_n)] = 1
         self.seen_coords_since_blackout.add((x_pos, y_pos, map_n))
-        self.explore_map[local_to_global(y_pos, x_pos, map_n)] = 1
+        try:
+            self.explore_map[local_to_global(y_pos, x_pos, map_n)] = 1
+        except IndexError:
+            gy, gx = local_to_global(y_pos, x_pos, map_n)
+            print(f"coord out of bounds! global: ({gx}, {gy}) game: ({x_pos}, {y_pos}, {map_n})")
```
Comment (thatguy11325) [inline pokemonred_puffer/environment.py]:
```suggestion
    self.explore_map[local_to_global(y_pos, x_pos, map_n)] = 1
```

looks like this is handled by the local_to_global function. But if you really want to do something like this, I'd change local_to_global to return an `Optional[tuple[int, int]]` and change this section to read like:

```python
if rc := local_to_global(y_pos, x_pos, map_n):
    self.explore_map[rc[0], rc[1]] = 1
```

File: pokemonred_puffer/environment.py (hunk 1 above)
Comment (kywch) [reply pokemonred_puffer/environment.py]: fixed!

File: pokemonred_puffer/global_map.py (hunk 2 above)
Comment (kywch) [reply pokemonred_puffer/global_map.py]: used 0, 0

File: pokemonred_puffer/environment.py (hunk 3 above)
Comment (kywch) [reply pokemonred_puffer/environment.py]: removed the comment

File: pokemonred_puffer/environment.py (hunk 4 above)
Comment (kywch) [reply pokemonred_puffer/environment.py]: This is my rage-edit. IndexError should be already handled by local_to_global

**Review by thatguy11325 — APPROVED** [review]

**Pull request merged**